### DISCUSSION TITLE
关于生成llvm-ir的行数

### DISCUSSION THREAD

**Discussion by nobodxbodon** [yuyan-lang/yuyan#75] [Q&A] [answered]
看到 [Numba](https://numba.readthedocs.io/en/stable/developer/architecture.html#compiler-stages) 的 add 函数生成的 ir 在[三十行左右](https://numba.readthedocs.io/en/stable/developer/architecture.html#stage-6a-generate-nopython-llvm-ir)，但之前 [加法](https://github.com/yuyan-lang/yuyan/discussions/73#discussioncomment-8184485) 生成了数百行 ir，不知两者是否性质有不同？

**Comment by UltimatePea** [comment, marked as answer]
算法很不一样，豫言还在早期版本，最终的beta版本，优化结果1+2可能会在20行内，但离那可能还要一年左右的时间。

**Reply by nobodxbodon** [reply]
是否可以使用现有语言或编译器后端，以便集中精力于设计与前端，待需要进一步优化时再改后端？

**Reply by UltimatePea** [reply]
是的，我们使用了LLVM的作为后端。

但目前生成的代码的执行速度非常慢，严重影响了编译器本身的开发，所以需要先集中精力攻克速度问题。我尝试过SML与JS作为目标语言，它们的执行速度与LLVM代码并无差别，所以目前决定仅支持LLVM。

**Reply by nobodxbodon** [reply]
> 我尝试过SML与JS作为目标语言

请问可以分享一下 js 为目标的实现吗？试过 py 吗？

**Reply by UltimatePea** [reply]
https://github.com/yuyan-lang/yuyan/pull/57
这个PR的最后应该是一个大概率能用的JS实现，用--target=js传参数

python本身就比javascript慢很多，所以没有考虑用python作为后端。

**Reply by nobodxbodon** [reply]
从 [此图](https://github.com/yuyan-lang/yuyan/blob/master/docs/result2.png) 看，各目标语言之间的差距似乎没那么大。如果做到和任一目标语言native代码的效率在一个数量级的话，就可以认为“性能够”了吧？

**Reply by UltimatePea** [reply]
其实只要和python一个速度就至少可以用了，现在还没发用

**Reply by nobodxbodon** [reply]
我试试把比如 quick-sort 手工编译成python看看效果如何。
